File: CodingInterview/CodingInterview/Strings/Longest_Palindromic_Substring.js
/*
5. Longest Palindromic Substring - LeetCode
Given a string s, find the longest palindromic substring in s.You may assume that the maximum length of s is 1000.

Example 1:
Input: "babad"
Output: "bab"
Note: "aba" is also a valid answer.

Example 2:
Input: "cbbd"
Output: "bb"
*/

/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {

    if (!s)
        return '';
    else if (s.length <= 1)
        return s;

    let word = '';
    let palindrome = '';

    for (let i = 0; i < s.length; i++) {

        for (let j = i + 1; j < s.length; j++) {
            word = s.substr(i, (j - i) + 1);

            if (IsPalindrome(word) && word.length > palindrome.length) {
                palindrome = word;
            }
        }

        if (palindrome.length === 0)
            palindrome = s[i];
    }

    return palindrome;
};

function IsPalindrome(s) {
    let result = '';
    for (let l of s) {
        result = l + result;
    }
    return result === s;
}


//BEST APPROACH ************************************************
var longestPalindrome = function (s) {

    let maxpalindrome = '';

    for (let i = 0; i < s.length; i++) {
        maxpalindrome = checkPalindrome(s, maxpalindrome, i, i);
        maxpalindrome = checkPalindrome(s, maxpalindrome, i, i + 1);
    }

    return maxpalindrome;
}

function checkPalindrome(s, maxpalindrome, left, right) {
    while (s[left] && s[left] === s[right]) {
        left--;
        right++;
    }

    if (right - left - 1 > maxpalindrome.length) {
        return s.substring(left + 1, right);
    }

    return maxpalindrome;
}

File: CodingInterview/CodingInterview/Strings/Longest_Palindromic_Substring.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CodingInterview.Strings
{
    class Longest_Palindromic_Substring
    {
        public string LongestPalindrome(string s)
        {
            string result = string.Empty;

            for (int i = 0; i < s.Length; i++)
            {

                string res1 = IsPalindrome(s, i, i);
                string res2 = IsPalindrome(s, i, i + 1);

                if (res1.Length > res2.Length && res1.Length > result.Length)
                    result = res1;
                else if (res2.Length > res1.Length && res2.Length > result.Length)
                    result = res2;
            }

            return result;
        }

        public string IsPalindrome(string s, int st, int ed)
        {

            if (s == null || st > ed) return null;

            while (st >= 0 && ed < s.Length && s[st] == s[ed])
            {
                st--;
                ed++;
            }

            return s.Substring(st + 1, ed - st - 1);

        }
    }
}
/*
 5. Longest Palindromic Substring
Medium

Given a string s, return the longest palindromic substring in s.

 

Example 1:

Input: s = "babad"
Output: "bab"
Note: "aba" is also a valid answer.
Example 2:

Input: s = "cbbd"
Output: "bb"
Example 3:

Input: s = "a"
Output: "a"
Example 4:

Input: s = "ac"
Output: "a"
 

Constraints:

1 <= s.length <= 1000
s consist of only digits and English letters (lower-case and/or upper-case),
 */